bliss/cached_dataset.py

```diff
@@ -122,26 +230,42 @@ def parse_slices(self, splits: str, length: int):
     def train_dataloader(self):
         assert self.file_paths[self.slices[0]], "No cached data found"
         transform = transforms.Compose(self.train_transforms)
-        my_dataset = MyIterableDataset(
+        dataset_type = MyIterableDataset if self.use_iterable_dataset else MyDataset
```

Comment:
Is it possible to always use `MyIterableDataset`, rather than having two types of datasets? It seems like `MyIterableDataset` is (or could become) flexible enough to handle all our use cases. If that's right, it'd be better not to have two ways to do it.

Reply:
Oh, is the issue distributed training? Can we generalize `MyIterableDataset` to support distributed training somehow? If so, we wouldn't have to maintain two distinct data loading pipelines.

Reply:
i'll try it

Reply:
I found a notebook that describes how to write distributed iterable dataset ([this link](https://colab.research.google.com/drive/1OFLZnX9y5QUFNONuvFsxOizq4M-tFvk-?usp=sharing)). It seems that it will treat iterable dataset as a map dataset, and the random shuffle is hard to realize. Another way is to rewrite our iterable dataset using `rank` information, but I suggest using PyTorch's `DistributedSampler` (in our code, it's `MyDistributedSampler`) so that it can automatically hanle the workers and distributed processes using PyTorch's API.

Reply:
This may be worth considering too:
https://pytorch.org/data/main/torchdata.datapipes.iter.html
It appears to be an alternative to an `IterableDataset`.
Not that we should be carried away with optimizing something that already works, but writing out tens of thousands of "split files" seems problematic: hard drives don't provide good throughput if data is accessed as small blocks in a random order.

Reply:
Thanks! I'll read this link. By the way, in our `MyDataset`, it reads the data by chunk (equivalent to `MyIterableDataset`) so we are now not using tens of thousands of "split files".

Reply:
yeah, it's equivalent to `MyIterableDataset`; i'll delete `MyIterableDataset`

Reply:
Great!